### PR TITLE
Rework `undocumented_unsafe_blocks`

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #8264
fixes: #8449

One thing came up while working on this. Currently comments on the same line are supported like so:

```rust
/* SAFETY: reason */ unsafe {}
```

Is this worth supporting at all? Anything other than a couple of words doesn't really fit well.

edit: [zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60undocumented_unsafe_blocks.60.20same.20line.20comment)

changelog: Don't lint `undocumented_unsafe_blocks` when the unsafe block comes from a proc-macro.
changelog: Don't lint `undocumented_unsafe_blocks` when the preceding line has a safety comment and the unsafe block is a sub-expression.

### CODE REVIEW

**Comment by Jarcho** [comment]
This should be good to go now.

File: tests/ui/undocumented_unsafe_blocks.rs [hunk 1]
```diff
@@ -89,11 +93,6 @@ fn block_comment_newlines() {
     unsafe {}
 }
 
-#[rustfmt::skip]
-fn inline_block_comment() {
-    /* Safety: */unsafe {}
```
Comment (giraffate) [inline tests/ui/undocumented_unsafe_blocks.rs:94]: It would be better to leave this as it is and add comments not to support this style.

**Review by giraffate — COMMENTED** [review]
Can the comment about style that are no longer supported be added to `What it does` section on this lint?

**Review by giraffate — APPROVED** [review]

**Comment by giraffate** [comment]
@bors r+

It looks good, thanks!

**Pull request merged**